gtsfm/densify/patchmatchnet_data.py

```diff
@@ -0,0 +1,267 @@
+                in self.keys, with 2 values indicating [min_depth, max_depth]
+        """
+        num_tracks = self._sfm_result.number_tracks()
+
+        # Initialize the pairwise scores between the same views as negative infinity
+        pair_scores = np.zeros((self._num_images, self._num_images))
+        np.fill_diagonal(pair_scores, -np.inf)
+
+        # Initialize empty lists to collect all possible depths for each view
+        depths: List[List[float]] = [[] for _ in range(self._num_images)]
+
+        for j in range(num_tracks):
+            track = self._sfm_result.get_track(j)
+            num_measurements = track.number_measurements()
+            measurements = [track.measurement(k) for k in range(num_measurements)]
+            w_x = track.point3()
+            for k1 in range(num_measurements):
+                for k2 in range(k1 + 1, num_measurements):
+                    i_a = measurements[k1][0]
+                    i_b = measurements[k2][0]
+
+                    key_a = -1
+                    key_b = -1
+
+                    # Check if measurement j1 belongs to a valid camera a
+                    if i_a in self._camera_idx_to_patchmatchnet_idx:
+                        key_a = self._camera_idx_to_patchmatchnet_idx[i_a]
+                        # Calculate track_i's depth in the camera pose
+                        a_z = self._sfm_result.get_camera(i_a).pose().transformTo(w_x)[-1]
+                        # Update image i_a's depth list only when i_a in a valid camera
+                        depths[key_a].append(a_z)
+
+                    # Check if measurement j2 belongs to a valid camera b
+                    if i_b in self._camera_idx_to_patchmatchnet_idx:
+                        key_b = self._camera_idx_to_patchmatchnet_idx[i_b]
+                        # Calculate track_i's depth in the camera pose
+                        b_z = self._sfm_result.get_camera(i_b).pose().transformTo(w_x)[-1]
+                        # Update image i_b's depth list only when i_b in a valid camera
+                        depths[key_b].append(b_z)
+
+                    # If both cameras are valid cameras
+                    if key_a >= 0 and key_b >= 0:
+                        # Calculate score for track_i in the pair views (cam_a, cam_b)
+                        score_a_b = piecewise_gaussian(
+                            xPa=self._camera_centers[i_a] - w_x, xPb=self._camera_centers[i_b] - w_x
+                        )
+                        # Sum up pair scores for each track_i
+                        pair_scores[key_a, key_b] += score_a_b
+                        pair_scores[key_b, key_a] += score_a_b
+
+        # Sort pair scores, for i-th row, choose the largest (num_views-1) scores, the corresponding views are selected
+        #   as (num_views-1) source views for i-th reference view.
+        pairs = np.argsort(pair_scores, axis=1)[:, -self._num_views + 1 :][:, ::-1]
```

Comment:
i'm a bit confused by all the operations here. Can we split this into two lines, so we only index once? If we want to sort from largest to smallest, I recommend `np.argsort(-pair_scores, axis=1`.